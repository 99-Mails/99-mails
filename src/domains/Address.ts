import type { Domain } from "./Domain";
import type { Mail } from "./Mail";
import type { ID } from "./Shared";

export type Address = {
  id: ID;

  /**
   * Email address string <login>@<domain>
   */
  address: string;

  /**
   * <domain> part of .address.
   * Note: it's not domain name string! It's `Domain` object!
   */
  domain: Domain;

  /**
   * Key that can be used to restore the access to this address after session terminated.
   * See MutationRoot.restoreAddress
   */
  restoreKey: string;

  /**
   * List of all emails received by this address.
   * It's recommended to use Session.mails instead
   */
  mails: [Mail];
};

export type AddressID = Pick<Address, "address" | "id">;
