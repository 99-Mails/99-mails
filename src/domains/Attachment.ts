export type Attachment = {
  /**
   * ID
   */
  id: string;

  /**
   * Raw attachment body can be downloaded by this URL
   */
  downloadUrl: URL;

  /**
   * Name of the attached file
   */
  name: string;

  /**
   * MIME-type of the attached file
   */
  mime: string;

  /**
   * Raw contents of the attachment (only when it's utf8 plaintext)
   */
  raw: string;

  /**
   * Size of the raw payload
   */
  rawSize: BigInt;
};
