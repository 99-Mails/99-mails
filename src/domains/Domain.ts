import type { Timestamp, AccessInterface } from "./Shared";

export type Domain = {
  /**
   * ID
   */
  id: string;
  /**
   * Fully-qualified domain name
   */
  name: string;
  /**
   * Timestamp of when this domain was first introduced in dropmail.me (how old is this domain)
   */
  introducedAt: Timestamp;
  /**
   * List of interfaces the domain is available through.
   * When domain is not available via some interface, it means `Address` can not be created or restored
   * via this interface.
   */
  availableVia: AccessInterface;
};
