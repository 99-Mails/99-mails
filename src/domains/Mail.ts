import type { DecodeStatus, TextSource, Timestamp } from "@/types";
import type { Attachment } from "./attachment";

export type Mail = {
  /**
   * ID
   */
  id: string;

  /**
   * Raw email (same as `raw` field) can be downloaded from this URL
   */
  downloadUrl: URL;

  /**
   * timestamp of when this email was received
   */
  receivedAt: Timestamp;

  /**
   * Raw unmodified value of the receiving email address string (so, when extended address was used,
   * this field will contain extended address)
   */
  toAddrOrig: string;

  /**
   * Value of MIME `From` header (only when decodeStatus is OK)
   */
  headerFrom: string;

  /**
   * Value of MIME `Subject` header (only when decodeStatus is OK)
   */
  headerSubject: string;

  /**
   * Cleaned text payload of the email. When MIME contains `text/plain` version, we put it here.
   * When MIME has only HTML body, we extract text from HTML (see textSource).
   * It's NULL when decodeStatus is ERROR_DECODING
   */
  text: string;

  /**
   * If MIME has HTML version, this field contains raw value of this HTML
   */
  html: string;

  /**
   * How `text` field was populated (from MIME `text/plain` or `text/html`)
   */
  textSource: TextSource;

  /**
   * Value of SMTP `RCPT TO` command
   */
  fromAddr: string;

  /**
   * Raw contents of the email exactly as it was received
   */
  raw: string;

  /**
   * Size of `raw` field
   */
  rawSize: number;

  /**
   * Reflects the extent to which this MIME mail was decoded (not all emails are valid MIME, so sometimes we are not able to decode them)
   */
  decodeStatus: DecodeStatus;

  /**
   * Just <login>@<domain> email address.
   * When extended address was used, extensions are dropped from this field
   */
  toAddr: string;

  /**
   * true if email has text/html body version.
   * false when there is no HTML version
   * null when `decodeStatus` is ERROR_DECODING.
   */
  hasHtml: boolean;

  /**
   * List of mail's attachents, if any and only when decodeStatus is STRICT
   */
  attachments: Attachment[];
};

// """
// If MIME has HTML version, this field contains sanitized version of this HTML:
// * `<script>` tags are cleaned
// * `<style>` tags are cleaned
// * `<link>` tags are removed
// * `<iframe>` tags are removed
// * `<meta>` tags are removed
// * `style` attributes are removed
// * `on*` attributes are removed
// * `href=\"javascript:\"` are removed
// * `<img src="">` is cleaned, based on (nullable) sanitizerConfig
// When `sanitizerConfig` is not provided, default `HtmlSanitizerConfig` is applied
// """
// sanitizedHtml(sanitizerConfig: HtmlSanitizerConfig): String

// """
// Similar to `sanitizedHtml`, but returns an URL where that same sanitized HTML can be downloaded.
// """
// downloadSanitizedHtmlUrl(sanitizerConfig: HtmlSanitizerConfig): Url

export type MailSummary = Pick<
  Mail,
  "headerSubject" | "text" | "fromAddr" | "downloadUrl" | "toAddr"
>;
