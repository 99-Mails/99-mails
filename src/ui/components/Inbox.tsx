import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Tag,
  Flex,
  Card,
  Link,
  Text,
  Center,
  Heading,
  Spinner,
  CardBody,
  CardHeader,
  CardFooter,
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
} from "@chakra-ui/react";
import { useEmail } from "../../services/api";
import { useStore } from "../../services/storeService";

const InboxHeadItem = ({ text, value }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

const InboxItem = ({ mail }) => {
  const { headerSubject, text, fromAddr, downloadUrl, rawSize, toAddr } = mail;

  return (
    <AccordionItem>
      {({ isExpanded }) => (
        <Card boxShadow="none" border="1px" borderColor="gray.200">
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem text="Received at:" value={"now"} />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
};

const InboxContainer = () => {
  const { sessionID } = useStore();

  const { fetchIncomingEmails } = useEmail();
  const { data, loading, error } = fetchIncomingEmails(sessionID);

  return (
    <>
      <Flex flexDirection={"column"} gap="4">
        <Heading size="xs">
          <Flex gap={2} alignItems={"center"}>
            <Text fontSize="3xl">Inbox</Text>
            <Tag bg="gray.500" color="white">
              {data?.session.mails.length}
            </Tag>
          </Flex>
        </Heading>
        {!data?.session.mails.length && (
          <Card bg="gray.100">
            <CardHeader>
              <Heading size="md">No Emails Yet!</Heading>
            </CardHeader>
          </Card>
        )}
        <Accordion allowMultiple>
          {loading && (
            <Center>
              <Spinner />
            </Center>
          )}
          {data &&
            data.session.mails.map((mail, index) => (
              <InboxItem key={index} mail={mail} />
            ))}
        </Accordion>
      </Flex>
    </>
  );
};

export { InboxItem };
export default InboxContainer;
