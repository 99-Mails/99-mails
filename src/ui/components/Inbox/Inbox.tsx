import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Tag,
  Flex,
  Card,
  Link,
  Text,
  Center,
  Heading,
  Spinner,
  CardBody,
  CardHeader,
  CardFooter,
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
} from "@chakra-ui/react";
import { useFetchIncomingEmails } from "../../../services/api";
import { useTempEmail } from "../../../services/tempEmailAdaptor";
import type { Mail } from "../../../domains/Mail";
import ErrorBoundary, { ErrorFallback } from "../ErrorBoundary";

const InboxHeadItem = ({ text, value }: { text: string; value: string }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

const InboxItem = ({ mail, ...props }: { mail: Mail }) => {
  const { headerSubject, text, fromAddr, downloadUrl, rawSize, toAddr } = mail;

  return (
    <AccordionItem {...props}>
      {({ isExpanded }) => (
        <Card boxShadow="none" border="1px" borderColor="gray.200">
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem text="Received at:" value={"now"} />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl.href} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
};

const InboxContainer = () => {
  const { sessionID } = useTempEmail();
  const { data, loading, error, refetch } = useFetchIncomingEmails(sessionID);

  return (
    <>
      <ErrorBoundary
        FallbackComponent={ErrorFallback}
        onReset={() => refetch()}
      >
        <Flex flexDirection={"column"} gap="4">
          <Heading size="xs">
            <Flex gap={2} alignItems={"center"}>
              <Text fontSize="3xl">Inbox</Text>
              <Tag bg="gray.500" color="white">
                {data?.session.mails.length}
              </Tag>
            </Flex>
          </Heading>
          {!data?.session.mails.length && (
            <Card bg="gray.100">
              <CardHeader>
                <Heading size="md">No Emails Yet!</Heading>
              </CardHeader>
            </Card>
          )}
          <Accordion allowMultiple>
            {error && (
              <Center>
                <Text>Error!</Text>
              </Center>
            )}
            {loading && (
              <Center>
                <Spinner />
              </Center>
            )}
            {data &&
              data.session.mails.map((mail: Mail, index: number) => (
                <InboxItem key={index} mail={mail} />
              ))}
          </Accordion>
        </Flex>
      </ErrorBoundary>
    </>
  );
};

export { InboxItem, InboxContainer };
