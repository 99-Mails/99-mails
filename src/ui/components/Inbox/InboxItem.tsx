import { AddIcon, MinusIcon, DownloadIcon } from "@chakra-ui/icons";
import {
  Flex,
  Card,
  Link,
  Text,
  CardBody,
  CardHeader,
  CardFooter,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
  AccordionItemProps,
  forwardRef,
} from "@chakra-ui/react";
import type { MailSummary } from "@/domains/Mail";
import { FancyDate } from "@/lib/date";
import { memo } from "react";

const InboxHeadItem = ({ text, value }: { text: string; value: string }) => {
  return (
    <Flex gap="4">
      <Text as="b">{text}</Text>
      <Text>{value}</Text>
    </Flex>
  );
};

type InboxItemProps = AccordionItemProps & {
  mail: MailSummary;
};

const InboxItem = memo(forwardRef<InboxItemProps, "div">((props, ref) => {
  const { mail, ...restProps } = props;
  const { headerSubject, toAddr, fromAddr, text, downloadUrl } = mail;

  return (
    <AccordionItem ref={ref} {...restProps} w="full" >
      {({ isExpanded }) => (
        <Card
          boxShadow="none"
          border="1px"
          borderColor="gray.200"
          borderRadius="0"
        >
          <h2>
            <AccordionButton>
              {isExpanded ? (
                <>
                  <CardHeader flex="1" flexDirection="row" textAlign="left">
                    <InboxHeadItem text="Subject:" value={headerSubject} />
                    <InboxHeadItem text="Recipient:" value={toAddr} />
                    <InboxHeadItem text="Sender:" value={fromAddr} />
                    <InboxHeadItem
                      text="Received at:"
                      value={new FancyDate().toISOString()}
                    />
                  </CardHeader>
                  <MinusIcon fontSize="12px" />
                </>
              ) : (
                <>
                  <CardHeader as="span" flex="1" textAlign="left">
                    <Text as="b">{headerSubject}</Text>
                  </CardHeader>
                  <AddIcon fontSize="12px" />
                </>
              )}
            </AccordionButton>
          </h2>
          <AccordionPanel pb={4}>
            <CardBody border="1px" borderColor="gray.300" p="3">
              {text}
            </CardBody>
            <CardFooter>
              <Link color="blue" href={downloadUrl.href} target="_blank">
                <Flex gap="2" alignItems="center">
                  <DownloadIcon />
                  Download Email
                </Flex>
              </Link>
            </CardFooter>
          </AccordionPanel>
        </Card>
      )}
    </AccordionItem>
  );
}));

export { InboxItem, InboxHeadItem };
