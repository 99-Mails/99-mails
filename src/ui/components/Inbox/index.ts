export * from "./Inbox";
export * from "./InboxItem";
