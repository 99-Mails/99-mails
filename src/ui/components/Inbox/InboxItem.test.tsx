import { it, expect } from "vitest";
import { MailSummary } from "@/domains/Mail";
import { render, screen } from "@/tests/test-utils";
import { InboxItem } from "./InboxItem";
import { InboxBody } from "./Inbox";

const downloadUrl: URL = new URL("http://downloadUrl.example.com/data.json");

const mail: MailSummary = {
  headerSubject: "My mail header",
  text: "My mail body",
  fromAddr: "admin@example.com",
  downloadUrl: downloadUrl,
  toAddr: "temp-email-address@my-mail-provider.com",
};

it("should render correctly", () => {
  const { asFragment } = render(
    <InboxBody>
      <InboxItem mail={mail} />
    </InboxBody>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("should have the correct body and subject-header", () => {
  render(
    <InboxBody data-testid="inbox-wrapper">
      <InboxItem mail={mail} />
    </InboxBody>
  );

  expect(screen.getByTestId("inbox-wrapper")).toHaveTextContent(
    "My mail header"
  );
  expect(screen.getByTestId("inbox-wrapper")).toHaveTextContent("My mail body");
});
