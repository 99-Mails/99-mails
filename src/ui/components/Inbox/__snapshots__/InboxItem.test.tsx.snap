// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-accordion css-0"
  >
    <div
      class="chakra-accordion__item css-1fsnuue"
    >
      <div
        class="chakra-card css-1p9ywwq"
      >
        <h2>
          <button
            aria-controls="accordion-panel-:r0:"
            aria-expanded="false"
            class="chakra-accordion__button css-ubg1nw"
            data-index="0"
            id="accordion-button-:r0:"
            type="button"
          >
            <span
              class="chakra-card__header css-1eziwv"
            >
              <b
                class="chakra-text css-0"
              >
                My mail header
              </b>
            </span>
            <svg
              class="chakra-icon css-1hiix9"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </h2>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: none; opacity: 0; height: 0px;"
        >
          <div
            aria-labelledby="accordion-button-:r0:"
            class="chakra-accordion__panel css-b0qdn7"
            id="accordion-panel-:r0:"
            role="region"
          >
            <div
              class="chakra-card__body css-18zq9bd"
            >
              My mail body
            </div>
            <div
              class="chakra-card__footer css-k008qs"
            >
              <a
                class="chakra-link css-14ksm7b"
                href="http://downloadurl.example.com/data.json"
                target="_blank"
              >
                <div
                  class="css-r2hvuq"
                >
                  <svg
                    class="chakra-icon css-onkibi"
                    focusable="false"
                    viewBox="0 0 14 14"
                  >
                    <path
                      d="M11.2857,6.05714 L10.08571,4.85714 L7.85714,7.14786 L7.85714,1 L6.14286,1 L6.14286,7.14786 L3.91429,4.85714 L2.71429,6.05714 L7,10.42857 L11.2857,6.05714 Z M1,11.2857 L1,13 L13,13 L13,11.2857 L1,11.2857 Z"
                      fill="currentColor"
                    />
                  </svg>
                  Download Email
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
